### PR TITLE
Model debug

### PR DESCRIPTION
**Pull request by YoojLee**
## Model gradient vanishing
gradient가 학습 초기에 제대로 발생하지 않으며, random pick에서 학습이 멈춰버리는 문제가 발생함.
image
image
원인은 Weight Decay가 optimizer와 맞지 않는 문제였음. (정확한 why는 아직 발견 못함) → AdamW로 optimizer 수정함

## Learning Rate Scheduling
cosine decay 중 learning rate 스케쥴링에 있어서 learning rate이 warmup 이후 constant로 멈춰버리는 문제 발생.  progress 업데이트에 문제가 있었던 것으로 확인.
cosine annealing 구현은 되어 있지 않아서, cosine annealing도 추가적으로 구현. cosine annealing에서도 learning rate이 멈춰버리는 문제가 발생함. 이 경우에는 주기 설정에 따라서 learning rate progress가 1에서 바뀌지 않아, cosine factor가 0으로 멈춰버리는 문제가 발생함.
image
 cycle_factor를 0으로 준 후 체크해보니 해당 문제가 발생하지 않아, period 업데이트에 있어서 발생한 문제로 파악.
period 업데이트 시 1+cycle_factor를 지속적으로 곱하면서 정수로 typecasting을 해주지 않은 문제였음. 수정 후, 정상 작동 확인 완료.
image

### CODE REVIEW

File: scheduler.py [hunk 1]
```diff
@@ -29,23 +29,56 @@ def lr_lambda(self, step):
...
         
-        #progress = min(1, max(0,(step - self.warmup_steps) / float(self.total_steps - self.warmup_steps))) # clipping
-        #return 0.5 * (1. + torch.cos(math.pi*progress)) # 이게 tensorflow 수식. 그래서 이거 어쩌란 거냐? vit-pytorch랑은 다른데..
+        # min_eta = 0, max_eta = initial_lr로 가정. (그렇게 되니까 lr_lambda를 그냥 쓰는 거겠지 -> lr_lambda 자체가 lambda function으로 초기 학습률값을 조작하는 방식으로 동작)
+        # 얘는 annealing이 아니라 그냥 전체 step을 하나의 주기로 보는 것 같음. annealing을 하나 더 구현해보자!
+        progress = min(1, max(0,(step - self.warmup_steps) / float(self.total_steps - self.warmup_steps))) # clipping (progress를 0과 1 사이로 강제)
+        return 0.5 * (1. + math.cos(math.pi*progress*float(self.cycle_factor)))
+
+
+class WarmupCosineAnnealing(LambdaLR):
```
Comment (YoojLee) [inline scheduler.py:57]: cosine annealing의 경우에는 period를 전체 step - warmup step으로 설정하면 cosine decay로도 가능.

File: scheduler.py [hunk 2]
```diff
@@ -29,23 +29,56 @@ def lr_lambda(self, step):
...
     def lr_lambda(self, step):
+        # warm-up 적용
         if step < self.warmup_steps:
             return float(step) / float(max(1, self.warmup_steps))
         
-        #progress = min(1, max(0,(step - self.warmup_steps) / float(self.total_steps - self.warmup_steps))) # clipping
-        #return 0.5 * (1. + torch.cos(math.pi*progress)) # 이게 tensorflow 수식. 그래서 이거 어쩌란 거냐? vit-pytorch랑은 다른데..
+        # min_eta = 0, max_eta = initial_lr로 가정. (그렇게 되니까 lr_lambda를 그냥 쓰는 거겠지 -> lr_lambda 자체가 lambda function으로 초기 학습률값을 조작하는 방식으로 동작)
+        # 얘는 annealing이 아니라 그냥 전체 step을 하나의 주기로 보는 것 같음. annealing을 하나 더 구현해보자!
+        progress = min(1, max(0,(step - self.warmup_steps) / float(self.total_steps - self.warmup_steps))) # clipping (progress를 0과 1 사이로 강제)
```
Comment (YoojLee) [inline scheduler.py:53]: 이 부분은 annealing에서는 어차피 t_cur를 따로 둬서 0에서부터 카운팅을 시작하기 때문에 annealing 내부에서는 warmup steps를 빼주는 부분을 구현하지 않아도 됨.
